File: 20.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (char i : s.toCharArray()) {
            switch (i) {
                case '(':
                case '[':
                case '{':
                    stack.push(i);
                    break;
                case ')':
                    if (stack.isEmpty() || stack.pop() != '(') {
                        return false;
                    }
                    break;
                case ']':
                    if (stack.isEmpty() || stack.pop() != '[') {
                        return false;
                    }
                    break;
                case '}':
                    if (stack.isEmpty() || stack.pop() != '{') {
                        return false;
                    }
                    break;
                default:
                    break;
            }
        }
        return stack.isEmpty();
    }
}

File: Python/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        parenthese_map = {")": "(", "}": "{", "]": "["}
        for i in s:
            match i:
                case "(" | "{" | "[":
                    stack.append(i)
                case ")" | "}" | "]":
                    if not stack:
                        return False
                    if stack[-1] != parenthese_map[i]:
                        return False
                    stack.pop()

        return True if not stack else False
